### DISCUSSION TITLE
Is this correct behavior for typing with generics?

### DISCUSSION THREAD

**Discussion by Naqu6** [microsoft/pyright#9231] [Q&A] [answered]
Hello! I think there may be a bug in pyright, but I also totally may just be misunderstanding the typing rules.

Reproductions:

```
from dataclasses import dataclass
from typing import Generic, TypeVar

@dataclass
class B:
    b : int

T = TypeVar('T', bound='B')

class C(Generic[T]):
    def foo(self, t: T):
        print(t.b)
    
    def bar(self, t: T):
        self.foo(t.b)

    def baz(self, t: T):
        self.foo(t)

```

I think there is a type error in this code - self.foo needs to be called with a `T`.
However, in `bar`, we call with a type that should be an int (`t.b`).

`pyright` does not catch an error and reports that this code is fine.

Is this a bug in pyright, or is there some typing rule I don't understand?

```
pyright==1.1.384
```

**Comment by erictraut** [comment, marked as answer]
Yes, that's a bug. Could you please create a bug report issue so it is tracked? Thanks!

**Reply by Naqu6** [reply]
https://github.com/microsoft/pyright/issues/9232

**Reply by Naqu6** [reply]
Thanks for the quick response! :)
